formats/json/commonMain/src/kotlinx/serialization/json/internal/StreamingJsonDecoder.kt

```diff
@@ -225,3 +228,16 @@ internal class StreamingJsonDecoder internal constructor(
         return enumDescriptor.getElementIndexOrThrow(decodeString())
     }
 }
+
+@OptIn(ExperimentalSerializationApi::class)
+@ExperimentalUnsignedTypes
+internal class JsonDecoderForUnsignedTypes(private val reader: JsonReader, json: Json) :
+    AbstractDecoder() {
+    override val serializersModule: SerializersModule = json.serializersModule
+    override fun decodeElementIndex(descriptor: SerialDescriptor): Int = error("unsupported")
+
+    override fun decodeInt(): Int = reader.takeString().toUInt().toInt()
+    override fun decodeLong(): Long = reader.takeString().toULong().toLong()
+    override fun decodeByte(): Byte = reader.takeString().toUByte().toByte()
+    override fun decodeShort(): Short = reader.takeString().toUShort().toShort()
```

Comment:
This code will throw `NumberFormatException` , it would be nice to throw more specific `SE` instead